### PR TITLE
Track best model in epochs

### PR DESCRIPTION
**Pull request by xamm**
Implements tracking model with best loss over epochs and saving the model to an attribute in the object if specified by the user.
Best model can be retrieved by calling `get_best()`, but will return `None` if `save_best` parameter is not set to `True` when calling `fit()`.

- if `save_best` is true save best model in synthesizer object attribute (`_best_model`)
- add `get_best()` function:
  - returns best_model or None if not saved
- combine CTGAN Discriminator and Generator loss for best loss decision
  - use absolute values as loss can be negative
- TVAE loss is used directly
- losses are detached from computation graph and cloned

### CODE REVIEW

**Comment by xamm** [comment]
My Idea would be to provide 2 classes:
1. a `Callback` class which solely specifies an abstract method that is supposed to return an object of type `CallbackInformation`
2. the `CallbackInformation` class for holding information calculated during the callback function

The user can define an own callback logic inheriting from the `Callback` class and implementing the `callback` method.
As it is not only a callback method but a class saving state over multiple calls to the `callback` method is possible.

Inheriting from the `CallbackInformation` class lets the user specify which additional information should be saved to the model during the fitting process. As the model does not need this information (besides from the `early_stop` attribute) it is not necessary to let the model class know which attributes this information holds.

From the model part is is solely needed to specify which information should be passed down to the used-defined `callback` method and what should be done with it is handled by the user yielding a high flexibility.

In the test a simple example is given for the possible usage:
The `Callback` class simply counts how often it has been called and after 2 calls / 2 epochs it tells the model to stop training.
Additional information (not needed for the model) is saved by creating a subclass of the `CallbackInformation` and saved during calls to the `callback` method.

Problem:
- to use the `@dataclass` attribute Python 3.7 is needed
  - if Python 3.6 must be supported than the `__init__` method must be implemented directly in the `CallbackInformation` class and in subclasses created by users

**Comment by davebulaval** [comment]
@katxiao reason this PR was close?

**Comment by tssbas** [comment]
> @katxiao reason this PR was close?

I would also like to know this. I think that callback functionality is essential for successful training. It would allow to do custom metric calculation during training and consequently even early stopping. Though this PR seems to provide the basic requirements I am not sure if it wouldn't be just better to covert everything into PyTorch Lightning and rely on all the supported callbacks provided by the framework. Perhaps @katxiao can comment on why PyTorch Lightning has not been used initially and what is planned for the future regarding callback functionality. Thanks!

**Comment by katxiao** [comment]
The PR was closed by the author, so I do not know the reason. Issue #149 is still in our backlog.
